src/connectors/azure_log.py

```diff
@@ -0,0 +1,333 @@
+                                  PROPERTIES_RISK_EVENT_TYPES, PROPERTIES_RISK_LEVEL_AGGREGATED, PROPERTIES_RISK_LEVEL_DURING_SIGNIN,
+                                  PROPERTIES_RISK_STATE, PROPERTIES_STATUS, PROPERTIES_TOKEN_ISSUER_TYPE, PROPERTIES_USER_DISPLAY_NAME,
+                                  PROPERTIES_USER_ID, PROPERTIES_USER_PRINCIPAL_NAME, RESOURCE_ID, RESULT_DESCRIPTION, RESULT_SIGNATURE,
+                                  RESULT_TYPE, TENANT_ID, EVENT_TIME, LOADED_ON)
+FROM (
+ SELECT $1, HASH($1), $1:Level::NUMBER, $1:callerIpAddress::STRING, $1:category::STRING, $1:correlationId::STRING, $1:durationMs,
+        $1:identity::STRING, $1:location::STRING, $1:operationName::STRING, $1:operationVersion::STRING, $1:properties::VARIANT,
+        $1:properties.appDisplayName::STRING, $1:properties.appId::STRING, $1:properties.appliedConditionalAccessPolicies::VARIANT,
+        $1:properties.authenticationMethodsUsed::VARIANT, $1:properties.authenticationProcessingDetails::VARIANT,
+        $1:properties.clientAppUsed::STRING, $1:properties.conditionalAccessStatus::STRING, $1:properties.createdDateTime::TIMESTAMP_LTZ,
+        $1:properties.deviceDetail::VARIANT, $1:properties.id::STRING, $1:properties.ipAddress::STRING, $1:properties.isInteractive::BOOLEAN,
+        $1:properties.location::VARIANT, $1:properties.mfaDetail::VARIANT, $1:properties.networkLocationDetails::VARIANT,
+        $1:properties.processingTimeInMilliseconds::NUMBER, $1:properties.resourceDisplayName::STRING, $1:properties.resourceId::STRING,
+        $1:properties.riskDetail::STRING, $1:properties.riskEventTypes::VARIANT, $1:properties.riskLevelAggregated::STRING,
+        $1:properties.riskLevelDuringSignIn::STRING, $1:properties.riskState::VARIANT, $1:properties.status::VARIANT,
+        $1:properties.tokenIssuerType::STRING, $1:properties.userDisplayName::STRING, $1:properties.userId::STRING,
+        $1:properties.userPrincipalName::STRING, $1:resourceId::STRING, $1:resultDescription::STRING, $1:resultSignature::STRING,
+        $1:resultType::STRING, $1:tenantId::STRING, $1:time::TIMESTAMP_LTZ, CURRENT_TIMESTAMP()
+FROM @DATA.{name}_STAGE
+)
+"""
+    }
+
+    db.create_pipe(
+        name=f"{name}_PIPE",
+        sql=pipe_sql[options['log_type']],
+        replace=True)
+
+    db.execute(f'ALTER PIPE DATA.{name}_PIPE SET PIPE_EXECUTION_PAUSED=true')
+    db.execute(f'GRANT OWNERSHIP ON PIPE DATA.{name}_PIPE TO ROLE {SA_ROLE}')
+
+    return {'newStage': 'finalized', 'newMessage': 'Table, Stage, and Pipe created'}
+
+
+def ingest(name, options):
+    block_blob_service = BlockBlobService(
+        account_name=options['storage_account'],
+        sas_token=options['sas_token'],
+        endpoint_suffix=options['suffix']
+    )
+    base_name = name.rstrip('CONNECTION')
+
+    db.execute(f"select SYSTEM$PIPE_FORCE_RESUME('DATA.{base_name}_PIPE');")
+
+    files = block_blob_service.list_blobs(options['blob_name'])
```

Comment:
can we put potential `KeyErrors` up top, so that if it errors, it does it before it tries to do anything?

Reply:
resolved